### PR TITLE
[yargs] support generics for command middlewares

### PR DESCRIPTION
**Pull request by loderunner**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - [Failing middleware typings](https://stackblitz.com/edit/node-mh477b?file=index.ts)
  - [Successful middleware typings](https://stackblitz.com/edit/node-juf6ys?file=index.ts)

### CODE REVIEW

File: types/yargs/v15/yargs-tests.ts [hunk 1]
```diff
@@ -305,8 +305,9 @@ function Argv$command() {
             (argv) => {
                 console.dir(argv.url);
             },
-            // middlewares
-            [],
+            [(argv) => {
+                console.log(argv.url);
```
Comment (forivall) [inline types/yargs/v15/yargs-tests.ts:309]: do a type assertion on `argv.url` instead - ie `// $ExpectType string`

Comment (loderunner) [reply types/yargs/v15/yargs-tests.ts:309]: Done.

Comment (loderunner) [reply types/yargs/v15/yargs-tests.ts]: Not done. 😅 

Still working on it.

Comment (loderunner) [reply types/yargs/v15/yargs-tests.ts]: I think we're good, this time.

**Review by forivall — APPROVED** [review]
Thanks!

File: types/yargs/index.d.ts [hunk 2]
```diff
@@ -148,15 +148,15 @@ declare namespace yargs {
             description: string,
             builder?: BuilderCallback<T, U>,
             handler?: (args: ArgumentsCamelCase<U>) => void | Promise<void>,
-            middlewares?: MiddlewareFunction[],
+            middlewares?: Array<MiddlewareFunction<U>>,
```
Comment (jakebailey) [inline types/yargs/index.d.ts:151]: I don't really understand why these are being changed to `Array<MiddlewareFunction<U>>` and not just `MiddlewareFunction<U>[]`, but, it's not semantically different, so...

**Review by jakebailey — APPROVED** [review]

File: types/yargs/index.d.ts (hunk 2 above)
Comment (loderunner) [reply types/yargs/index.d.ts]: Using `MiddlewareFunction<U>[]`, I get a tslint error for `array-type`.

```
Array type using 'T[]' is forbidden for non-simple types. Use 'Array<T>' instead.
```

[tslint docs](https://palantir.github.io/tslint/rules/array-type/)

**Comment by loderunner** [comment]
Ready to merge

File: types/yargs/index.d.ts (hunk 2 above)
Comment (jakebailey) [reply types/yargs/index.d.ts]: Huh, TIL.

Comment (loderunner) [reply types/yargs/index.d.ts:151]: ... and I just pushed the change by mistake. 🤦‍♂️

**Review by jakebailey — APPROVED** [review]

**Comment by loderunner** [comment]
Ready to merge

**Pull request merged**